returnn/tf/layers/basic.py

```diff
@@ -859,70 +859,82 @@ class SliceNdLayer(_ConcatInputLayer):
     if size is None:
       if min_size is None:
         min_size = 0
-      if seq_lens is None:
+      if seq_lens_data is None:
         assert isinstance(x.batch_shape[x.time_dim_axis], int)
-        size = tf.maximum(tf.reduce_max(x.batch_shape[x.time_dim_axis] - start_t), min_size)  # scalar
+        size_t = x.batch_shape[x.time_dim_axis] - start_t
       else:
-        # make seq_lens compatible with start_t
-        seq_lens = expand_multiple_dims(  # e.g. (B,) or (B,1)
-          x=seq_lens,
-          axes=[-1] * (len(start_t.shape) - len(seq_lens.shape)))
-        size = tf.maximum(tf.reduce_max(seq_lens - start_t), min_size)  # scalar
+        seq_lens_t = seq_lens_data.copy_compatible_to(start_data, check_sparse=False).get_placeholder_as_batch_major()
+        size_t = seq_lens_t - start_t
+      size = tf.maximum(tf.reduce_max(size_t), min_size)  # scalar
+    elif isinstance(size, LayerBase):
+      size_data = size.output.copy()
+      common_data = Data.get_common_data([start_data, size_data])
+      size_data = size_data.copy_compatible_to(common_data)
+      size_t = size_data.get_placeholder_as_batch_major()
+      start_data = start_data.copy_compatible_to(common_data)
+      start_t = start_data.get_placeholder_as_batch_major()
+      min_size = 0
+      size = tf.maximum(tf.reduce_max(size_t), min_size)  # scalar
+    else:
+      # in this case, size_t is never used but needs to be set to avoid a PyCharm warning
+      size_t = None
     # for each start index in start_data, we want to gather a slice
     # therefore, the output's first axes are the same as the ones from start_data
     # and the next axis will therefore be the slice axis
     slice_tag = self.output.dim_tags[start_data.batch_ndim]
     assert slice_tag.description.startswith("sliced-time:")
     if not isinstance(size, int):
       # in this case, size is not known before runtime and becomes dynamic and we need to set dyn_size
-      if seq_lens is None:
-        dyn_size = tf.maximum(x.batch_shape[x.time_dim_axis] - start_t, min_size)  # (B,) or (B,T)
-      else:
-        dyn_size = tf.maximum(seq_lens - start_t, min_size)  # (B,) or (B,T)
+      dyn_size = tf.maximum(size_t, min_size)  # (B,) or (B,T)
       dyn_size_ext = Data(
         name=("%s:dyn_size" % slice_tag.description),
         dtype=Data.size_dtype,
         placeholder=dyn_size,
-        dim_tags=start_data.dim_tags,
+        dim_tags=start_data.copy_as_batch_major().dim_tags,  # as batch major because the placeholder is too
         batch=slice_tag.batch,
         beam=slice_tag.batch.beam if slice_tag.batch else self.output.beam,
         control_flow_ctx=slice_tag.control_flow_ctx)
       slice_tag.dyn_size_ext = dyn_size_ext
       slice_tag.set_tag_on_size_tensor(dyn_size)
-    gather_positions_data = start_data.copy_template(name="%s_gather_positions" % self.name)
+    # get as batch major, because the placeholder later will also be batch major
```

Comment:
Why is the placeholder batch major? The placeholder is the same as `start` (`start_t`), or not?